drivers/gpio/gpio_pca95xx.c

```diff
@@ -440,13 +454,146 @@ static int gpio_pca95xx_port_toggle_bits(struct device *dev, uint32_t mask)
 }
 
+#ifdef CONFIG_PCA95XX_INTERRUPT_SUPPORT
+static void gpio_pca95xx_work_handler(struct k_work *work)
+{
+	struct gpio_pca95xx_drv_data *ddata =
+			CONTAINER_OF(work, struct gpio_pca95xx_drv_data, work);
+	uint16_t status;
+	int i = 0;
+
+	if (read_port_regs(ddata->exp_dev, REG_INT_STATUS_PORT0,
+			   &status) != 0) {
+		return;
+	}
+
+	while (status) {
+		if (status & 0x1) {
+			gpio_fire_callbacks(&ddata->cb_list,
+					    ddata->exp_dev,
+					    BIT(i));
+		}
+		status >>= 1;
+		i++;
+	}
+}
+
+static void gpio_pca95xx_global_callback(struct device *port,
+					 struct gpio_callback *cb, uint32_t pin)
+{
+	struct gpio_pca95xx_drv_data *ddata =
+			CONTAINER_OF(cb, struct gpio_pca95xx_drv_data, gpio_cb);
+
+	ARG_UNUSED(port);
+	ARG_UNUSED(pin);
+
+	k_work_submit(&ddata->work);
+}
+
+static int gpio_pca95xx_interrupt_init(struct device *dev)
+{
+	struct gpio_pca95xx_drv_data *ddata = dev->data;
+	const struct gpio_pca95xx_config *cfg = dev->config;
+	int ret;
+
+	ddata->gpio = device_get_binding(cfg->gpio_port);
+	if (!ddata->gpio) {
+		LOG_ERR("Gpio controller %s not found.", cfg->gpio_port);
+		return -EINVAL;
+	}
+
+	ddata->work.handler = gpio_pca95xx_work_handler;
+
+	ret = gpio_pin_configure(ddata->gpio, cfg->int_pin,
+				 GPIO_INPUT | cfg->int_flags);
+	if (ret) {
+		LOG_ERR("gpio %d input config failed err %d.",
+			 cfg->int_pin, ret);
+		return ret;
+	}
+
+	gpio_init_callback(&ddata->gpio_cb, gpio_pca95xx_global_callback,
+			   BIT(cfg->int_pin));
+
+	ret = gpio_add_callback(ddata->gpio, &ddata->gpio_cb);
+	if (ret) {
+		LOG_ERR("gpio add callback err %d.", ret);
+		return ret;
+	}
+
+	ret = gpio_pin_interrupt_configure(ddata->gpio, cfg->int_pin,
+					   GPIO_INT_EDGE_FALLING);
+
+	return ret;
+}
+
+static int gpio_pca95xxx_manage_callback(struct device *dev,
+					 struct gpio_callback *cb,
+					 bool set)
+{
+	struct gpio_pca95xx_drv_data *ddata = dev->data;
+
+	return gpio_manage_callback(&ddata->cb_list, cb, set);
+}
+
 static int gpio_pca95xx_pin_interrupt_configure(struct device *dev,
-						  gpio_pin_t pin,
-						  enum gpio_int_mode mode,
-						  enum gpio_int_trig trig)
+			gpio_pin_t pin, enum gpio_int_mode mode,
+			enum gpio_int_trig trig)
 {
+	struct gpio_pca95xx_drv_data * const drv_data =
+		(struct gpio_pca95xx_drv_data * const)dev->data;
+	int ret = 0;
+
+	/* Level trigger interrupts not supported */
+	if (mode == GPIO_INT_MODE_LEVEL) {
+		return -ENOTSUP;
+	}
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	if (mode == GPIO_INT_MODE_DISABLED) {
+		if (!(drv_data->reg_cache.int_mask & ~BIT(pin))) {
+			ret = write_port_regs(dev, REG_INT_MASK_PORT0,
+				&drv_data->reg_cache.int_mask,
+				drv_data->reg_cache.int_mask | BIT(pin));
+
+			write_port_regs(dev, REG_POL_INV_PORT0,
+				&drv_data->reg_cache.pol_invert,
+				drv_data->reg_cache.pol_invert & ~BIT(pin));
+		}
+		goto exit;
+	}
+
+	ret = setup_pin_dir(dev, pin, GPIO_INPUT);
+	if (ret) {
+		goto exit;
+	}
+
+	if (trig & GPIO_INT_TRIG_LOW) {
```

Comment:
This bothers me.  I assume the intent is to support active-high interrupts, but the Zephyr GPIO driver infrastructure manages polarity in the common code base.  Certainly setting this would change cause the pin value read back through `gpio_pin_get()` to be inverted too, which is not going to be expected.